keps/sig-network/0031-20181109-multicast-support.md

```diff
@@ -0,0 +1,191 @@
+just accidentally as a consequence of IP/MAC filtering).
+
+Weave treats multicast as broadcast. There has been discussion of
+[implementing IGMP
+snooping](https://github.com/weaveworks/weave/issues/178) so that
+multicast packets would only be sent to the pods that want them, but
+this is not yet implemented. There has been some discussion of
+[multicast between pods and external
+IPs](https://github.com/weaveworks/weave/issues/1863), which
+apparently can be made to work at least in some circumstances. There
+is no support for policy/isolation; all multicast packets are sent to
+all pods in the cluster.
+
+OpenShift SDN also implements multicast-as-broadcast, but on a
+per-namespace basis, [with namespaces having to opt
+in](https://docs.okd.io/latest/admin_guide/managing_networking.html#admin-guide-networking-multicast).
+(So if you enable multicast on namespace "foo", and a pod in "foo"
+sends out a multicast packet, it will be delivered to all other pods
+in "foo".) As with Weave, it's not possible to use NetworkPolicy to
+restrict multicast traffic, and it's also not possible to use
+NetworkPolicy to extend multicast traffic across namespace boundaries.
+
+## Proposal
+
+### User Stories
+
+TBD
+
+### Implementation Details/Notes/Constraints
+
+#### Baseline Behavior: Block IP Multicast
+
+Although the Kubernetes network is normally open-by-default, except
+where NetworkPolicy dictates otherwise, it seems like
+closed-by-default is a better default for multicast:
+
+1. In many clusters, open-by-default multicast would be useful only as
+a denial of service attack vector.
+
+2. Even in cases where some pods in different namespaces on different
+nodes need to communicate via multicast, it is still likely to be the
+case that *most* pods will not be interested in multicast traffic, and
+sending the traffic to them would be a waste of bandwidth.
+
+3. This is probably the most-commonly-implemented behavior among
+existing network plugins anyway.
+
+#### Optional IP Multicast Support
+
+Plugins that want to provide support for IP multicast can allow users
+to configure it as described below. In that case, there are two levels
+of multicast support that the plugin might implement:
+
+1. For basic multicast support, a plugin can simply treat multicast as
+broadcast; when a pod sends a multicast packet, it will be transmitted
+to all other allowable recipients, regardless of whether those
+recipients have subscribed to any multicast groups. This is sufficient
+for at least basic use cases.
+
+2. For more advanced multicast support, a plugin could implement IGMP
+snooping, and monitor which multicast addresses each pod is interested
+in traffic for, and then only deliver multicast traffic to the pods
+which have subscribed to that traffic. This might be more efficient if
+there are many pods communicating over different multicast addresses.
+
+#### End-User Configuration of Multicast
+
+TBD
```

Comment:
I think we really need to see something here.  How is it configured?  What is the pod-level UX?  How does a user or pod know if it can rely on multicast?